relay-server/src/services/project_cache.rs

```diff
@@ -838,38 +847,69 @@ impl ProjectCacheBroker {
     ///
     /// The flushing of the buffered envelopes happens in `update_state`.
     fn handle_validate_envelope(&mut self, message: ValidateEnvelope) {
-        let ValidateEnvelope { envelope: context } = message;
+        let ValidateEnvelope {
+            envelope: mut managed_envelope,
+        } = message;
+
         let project_cache = self.services.project_cache.clone();
-        let envelope = context.envelope();
+        let envelope = managed_envelope.envelope();
 
         // Fetch the project state for our key and make sure it's not invalid.
         let own_key = envelope.meta().public_key();
-        let project_state = self
-            .get_or_create_project(own_key)
-            .get_cached_state(project_cache.clone(), envelope.meta().no_cache())
-            .filter(|st| !st.invalid());
+        let project = self.get_or_create_project(own_key);
+
+        let project_state =
+            project.get_cached_state(project_cache.clone(), envelope.meta().no_cache());
+
+        let project_state = match project_state {
+            ProjectState::Enabled(state) => Some(state),
+            ProjectState::Disabled => {
+                managed_envelope.reject(Outcome::Invalid(DiscardReason::ProjectId));
+                return;
+            }
+            ProjectState::Pending => None,
+        };
 
         // Also, fetch the project state for sampling key and make sure it's not invalid.
-        let sampling_key = utils::get_sampling_key(envelope);
-        let sampling_state = sampling_key.and_then(|key| {
-            self.get_or_create_project(key)
-                .get_cached_state(project_cache, envelope.meta().no_cache())
-                .filter(|st| !st.invalid())
-        });
+        let sampling_key = envelope.sampling_key();
+        let mut requires_sampling_state = sampling_key.is_some();
+        let sampling_state = if let Some(sampling_key) = sampling_key {
+            let state = self
+                .get_or_create_project(sampling_key)
+                .get_cached_state(project_cache, envelope.meta().no_cache());
+            match state {
+                ProjectState::Enabled(state) => Some(state),
+                ProjectState::Disabled => {
+                    relay_log::trace!("Sampling state is disabled ({sampling_key})");
+                    // We accept events even if its root project has been disabled.
+                    requires_sampling_state = false;
```

Comment:
This was missing in the previous PR: By setting `state` to `None`, we fell back to enqueueing an envelope if its sampling project was disabled (see `test_root_project_disabled`).